GFF Gurage Zebidar (Experimental) Keyboard v0.6

dyacob opened the pull request:
A keyboard, including  font, to support an experimental orthography for Gurage (Sebat Bet, SGW).

keyman-server commented:
This pull request is from an external repo and will not automatically be built. The build must still be passed before it can be merged. Ask one of the team members to make a manual build of this PR.

darcywong00 commented:
You'll need an `experimental/gff/gff_gurage/gff_gurage.keyboard_info` file. Something like
```
{
    "license": "mit",
    "languages": {
        "sgw-Ethi": {
            "font": {
              "family": "Zebidar",
              "source": [
                "Zebidar-R.ttf"
              ]
            }
        }
    },
    "description": "This is a Gurage (ጉራጌ , ISO-639-2 swg) language mnemonic input method."
}
```

darcywong00 commented on experimental/gff/gff_gurage/source/gff_gurage.kmn:
```diff
@@ -0,0 +1,241 @@
+c Other Info    :  http://keyboards.ethiopic.org/ , http://unicode.org/charts/
+c 
+  store(&Version) '9.0'
+store(&KEYBOARDVERSION) '0.6'
```
The PR and HISTORY.md file say version `0.5`

dyacob replied:
after reviewing the development history, i see that 0.6 is appropriate and updates have been committed now accordingly.

darcywong00 commented on experimental/gff/gff_gurage/README.md:
```diff
@@ -0,0 +1,23 @@
+
+Supported Platforms
+-------------------
+ * Windows
```
The targets in gff_gurage.kmn also include macOS and Linux

dyacob replied:
added.

darcywong00 commented on experimental/gff/gff_gurage/source/gff_gurage_and_amharic.kps:
```diff
@@ -0,0 +1,178 @@
+    </Keyboard>
+  </Keyboards>
+  <Strings/>
+</Package>
```
I wonder if it would be better to bundle the keyboards in `release/packages/` and have experimental/gff/gff_gurage only build gff_gurage.kmp

dyacob replied:
that works for me, whatever makes sense for the "experimental" folder policy.  my main concern is that when someone installs the keyboard (by whatever means), the installer knows to retrieve the corresponding font.

darcywong00 replied:
> whatever makes sense for the "experimental" folder

The `experimental` folder  is for keyboards that are under active development, but don't have all the files necessary for release, or are being tested by user community members before finalizing the design.


> my main concern is that ... the installer knows to retrieve the corresponding font.

As long as you've associated the fonts for each keyboard in the "Package Editor" tab of Developer, you should be fine.

Should you end up moving `gff_gurage_and_amharic.kps` to `release/packages/`, you can refer to #959 to see how to ensure keyboards get built before bundling together in `release/packages/`.

darcywong00 commented on experimental/gff/gff_gurage/gff_gurage.keyboard_info:
```diff
@@ -0,0 +1,16 @@
+      "font": {
+        "family": "Zebidar",
+        "source": [
+          "Zebidar-R.ttf",
```
Because the build script JSON validator is strict, you'll need to remove the trailing commas here and line 15.

dyacob replied:
done, sorry i missed that.

dyacob commented:
Is the recommended next step to move all contents to the "`experimental`" to "`release`" folder, and move `gff_gurage_and_amharic.kps` to `release/packages/` ?

darcywong00 commented:
> Is the recommended next step to move all contents to the "`experimental`" to "`release`" folder

That's more up to @DavidLRowe or @LornaSIL . fwiw, you've already got sufficient documentation for this keyboard

> and move `gff_gurage_andamharic.kps` to `release/packages/`?

Yes, that would be good practice.

LornaSIL commented:
@dyacob since @darcywong00 thinks you have sufficient documentation to move the keyboard from experimental to release, I'm all for moving your keyboard to release. I think you consider this an "experimental" orthography, but not an experimental keyboard, is that true?

Once you move the keyboard and package to release I'll start checking it.

dyacob commented:
@LornaSIL , yes the orthography is the experimental aspect, not the keyboard.  I've relocated the project folder under `release` and moved the gff_gurage_and_amharic.kps` into the `packages` folder and adjusted paths accordingly.   thanks!

LornaSIL commented:
Both of the pdfs have a title at the top for Tigrigna. It took me awhile to figure out where it was coming from, but the .doc files can be changed by opening the doc, going to File / Info and then clicking on the Title and changing it.
![image](https://user-images.githubusercontent.com/7071907/72292879-e73e7800-3617-11ea-8a5b-135cf979dded.png)

LornaSIL commented:
So, I think you should remove from the gff_gurage.kpj ALL the files mentioning Amharic.

Also, MOVE the file gff_gurage_and_amharic.kps to packages.

Look at the `GitHub\keyboards\release\packages\gff_amh_powerpack_7` and see how the package files are done. (Actually...why don't you just work on gff_gurage and finish it up and then do the package as a separate PR. That might make it less complicated.)

Also, I noticed that this PR makes a change to `/gff_amharic/source/gff_amharic.kps`

I think that change needs reverting since we don't want to modify the gff_amharic keyboard without bumping the version.

dyacob commented:
Thank for catching the incorrect titles -good eye :-)

dyacob commented:
Updates from the last few commits:

1. `gff_gurage_and_amharic.kps` moved into `packages` under a folder of the same name.
2.  `gff_amharic.kps` conflict resolved (should no longer be listed in the PR)
3. `gff_gurage.kpj` builds the `gff_gurage` package only (Amharic resources removed).

LornaSIL commented:
It seems like you are "hacking" the U+2DD0..U+2DD6 codepoints. I would have thought it would be better to put those at PUA codepoints. Are you planning to propose them to Unicode?

At any rate, I don't see those mentioned in the Zebidar FONTLOG.txt.

Another argument in favor of not hacking those codepoints is that you are going to package this keyboard with another keyboard (I do realize that Amharic doesn't use these codepoints either, so it may be a weak argument, but I REALLY dislike the idea of hacking Unicode if PUA works).

I didn't test every key sequence, but I tested keystrokes in every series. 

In keyboard_info, please delete the comma at the end of this line:
```
          "Zebidar-R.ttf",
```

Regarding the packages.

You'll need to add:
```
packages\gff_gurage_and_amharic\gff_gurage_and_amharic.keyboard_info
packages\gff_gurage_and_amharic\README.md
packages\gff_gurage_and_amharic\gff_gurage_and_amharic.kpj
```

You can look at the other gff package to help you if you aren't sure how to handle something.
Your keyboard_info file could be:
```
{
  "name": "Gurage and Amharic Keyboards",
  "license": "mit",
  "languages": {
    "sgw-Ethi": {
      "font": {
        "family": "Zebidar",
        "source": [
          "Zebidar-R.ttf"
        ]
      }
    },
    "am": {
      "font": {
        "family": "GeezWeb",
        "source": [
          "wookianos.woff",
          "wookianos.ttf",
          "wookianos.mobileconfig",
          "ETHIOPI0.eot"
        ]
      },
      "example": {
        "keys": "Tiena ysTlN",
        "text": "\u1324\u1293 \u12ED\u1235\u1325\u120D\u129D",
        "note": "\"Hello\""
      }
    }
  },
  "description": "This package provides both a Gurage and Amharic keyboard..."
}
```

dyacob commented:
Briefly on this topic:

> It seems like you are "hacking" the U+2DD0..U+2DD6 codepoints. I would have thought it would 
> be better to put those at PUA codepoints. Are you planning to propose them to Unicode?

The changes here should be viewed as a difference in glyph preferences only, vs the introduction of new letters for new sounds.   If the orthography sees bona fide adoption at the regional level, and the  Ethiopian Standards Agency accepts them at the national level, then a recommendation to change the  reference glyphs would follow from the ESA.

LornaSIL commented:
> 
> 
> Briefly on this topic:
> 
> > It seems like you are "hacking" the U+2DD0..U+2DD6 codepoints. I would have thought it would
> > be better to put those at PUA codepoints. Are you planning to propose them to Unicode?
> 
> The changes here should be viewed as a difference in glyph preferences only, vs the introduction of new letters for new sounds. If the orthography sees bona fide adoption at the regional level, and the Ethiopian Standards Agency accepts them at the national level, then a recommendation to change the reference glyphs would follow from the ESA.

I can accept that. Want to add that to the FONTLOG.txt?

dyacob commented:
FONTLOG.txt fixed, looks like it was a copy & paste style  error  and lines 17-23 were references  glyphs 12D0-12D6 which should have been 22D0-22D6.  Thanks for  catching that :-)

LornaSIL commented:
@dyacob the packages part of this PR is not right. What would you think about deleting the packages part from this PR and let me submit the packages in a separate PR? Or you can? That way I can go ahead and approve just the gurage keyboard.

If you want to do it, this is what I *think* needs to happen.
* Delete the .kps file in the source folder.
* Move the .kps file in the root to the source folder.
* Change the path in the .kpj file to include `source\`
* Remove the other .kps files from the .kpj

I haven't tested this yet, but I think that's what has to happen.

dyacob commented:
I think I have it fixed... but then I totally missed that a duplicate .kps file had crept in, that's a mystery.  The various "Build" buttons in KM Developer appear to be working correctly, no errors on my end.  If there is still a significant issue then I'll give up and go with the 2 separate PRs as you recommend.  thanks!

LornaSIL commented:
Ok, it is building now. Technically the gff_gurage.kvks does not need to be included in the package since it is a source file. You do include the .kvk file.

Otherwise, it looks good.

LornaSIL commented:
Sorry, but the online build failed. I'm pretty sure that the problem is that we need to get the keyboard on the build system first, then the packages will build.

* So, remove gff_gurage_and_amharic from the packages folder (save it on your system!) 

After I approve the keyboard and it gets merged, THEN you can submit the gff_gurage_and_amharic package in another PR.

I hope I'm right. I have a vague memory of this being an issue when we did the other packages.

dyacob commented:
oh dear, could it have failed because I pushed in an update while it was building?

LornaSIL commented:
>oh dear, could it have failed because I pushed in an update while it was building?

I don't think so, I think the package needs the compiled gff_gurage files and those may or may not have already been built when it started processing the package. That's why the keyboard needs approving first. I think that's true.

LornaSIL commented:
Ok, the keyboard built and is online. https://keyman.com/keyboards/gff_gurage
Please update your master
```
git checkout master
git fetch upstream
git merge upstream/master
git push origin master
```
BEFORE you submit the new package files.
